analyzers/tests/SonarAnalyzer.Test/TestCases/AspNet/AvoidUnderPosting.cs

```diff
@@ -0,0 +1,252 @@
+    public class DecoratedWithApiControlerAttribute : ControllerBase
+    {
+        [HttpGet]
+        public int Single(SingleModel model) => 42;
+
+        [HttpGet]
+        [HttpPost]
+        [HttpPut]
+        [HttpDelete]
+        public int Multiple(MultipleModel model) => 42;
+
+        [AcceptVerbs("POST")]
+        public int Verb(VerbModel model) => 42;
+
+        [AcceptVerbs("GET", "POST", "PUT", "DELETE")]
+        public int MultipleVerbs(MultipleVerbsModel model) => 42;
+    }
+}
+
+namespace ModelHierarchy
+{
+    public class Parent
+    {
+        public int ParentProperty { get; set; }         // Noncompliant
+    }
+
+    public class Child : Parent
+    {
+        public int ChildProperty { get; set; }          // Noncompliant
+    }
+
+    public class GrandChild : Child
+    {
+        public int GrandChildProperty { get; set; }     // Noncompliant
+    }
+
+    public class ControllerClass : Controller
+    {
+        [HttpPost] public IActionResult Create(GrandChild model) => View(model);
+        [HttpPost] public IActionResult Update(Parent model) => View(model);
+    }
+}
+
+namespace CompositeModels
+{
+    public class Model
+    {
+        public int Property { get; set; }               // Noncompliant
+        public Model SameModel { get; set; }
+        public AnotherModel AnotherModel { get; set; }
+    }
+
+    public class AnotherModel
+    {
+        public int AnotherProperty { get; set; }        // Noncompliant
+        public Model Model { get; set; }
+    }
+
+    public class ControllerClass : Controller
+    {
+        [HttpPost] public IActionResult Create(Model model) => View(model);
+    }
+}
+
+namespace GenericCollections
+{
+    public class EnumerableItem { public int Property { get; set; } }           // Noncompliant
+    public class ListItem { public int Property { get; set; } }                 // Noncompliant
+    public class DictionaryKeyItem { public int Property { get; set; } }        // Noncompliant
+    public class DictionaryValueItem { public int Property { get; set; } }      // Noncompliant
+
+    public class ControllerClass : Controller
+    {
+        [HttpPost] public IActionResult CreateEnumerable(IEnumerable<EnumerableItem> model) => View(model);
+        [HttpPost] public IActionResult CreateList(List<ListItem> model) => View(model);
+        [HttpPost] public IActionResult CreateDictionary(Dictionary<DictionaryKeyItem, DictionaryValueItem> model) => View(model);
+    }
+}
+
+namespace GenericModelWithTypeConstraint
+{
+    public class MyController : ControllerBase
+    {
+        public void Create(Model<Developer> model)
+        {
+        }
+    }
+
+    public class Model<T> where T : Person
+    {
+        public T Person { get; set; }
+    }
+
+    public abstract class Person
+    {
+        public int Age { get; set; }            // Noncompliant
+    }
+
+    public class Developer : Person
+    {
+        public string ProgramingLanguage { get; set; }
```

Comment:
Do we also get a `Noncompliant` in this class?

Reply:
No, the `ProgrammingLanguage` property is a nullable reference type, so the rule won't raise. For non-nullable reference types see[ this test case](https://github.com/SonarSource/sonar-dotnet/blob/4bf891effc509550499a05214dcf163c2f668ec1/analyzers/tests/SonarAnalyzer.Test/TestCases/AspNet/AvoidUnderPosting.CSharp8.cs#L10).

Reply:
Please add a test. It is not clear of the type argument is based on the constraint (Person) or the closing type (Developer). We need to make sure Developer is visited.